### PR TITLE
Add synchronous runtime and make it the default for Wasm system APIs

### PR DESCRIPTION
**Pull request by ma2bd**
## Motivation

* Use one thread for all contract calls of a transaction (#1193)
* One step closer to share memory states between distinct calls of the same application in the same transaction (#488)

## Proposal

* Create a new "synchronous" implementation of the system API traits
* Improve error handling
* Make the new runtime the default (but do not remove the old one for now)

## Test Plan

CI

### CODE REVIEW

File: linera-execution/src/execution.rs [hunk 1]
```diff
@@ -231,27 +271,79 @@ where
...
+                chain_id,
+                runtime_limits,
+                initial_remaining_fuel,
+                action,
+            )
+        });
+        while let Some(request) = execution_state_receiver.next().await {
+            self.handle_request(request).await?;
+        }
+        let (execution_results, runtime_counts) = execution_results_future.await??;
```
Comment (andresilva91) [inline linera-execution/src/execution.rs:326]: Rust n00b question: what do the double question mark `??` do here? I thought it was for when we had two `Result` types within another, but that doesn't seem to be the case here 🤔

Comment (ma2bd) [reply linera-execution/src/execution.rs:326]: There are two results to unpack. One from the closure and one from spawn_blocking

Comment (andresilva91) [reply linera-execution/src/execution.rs:326]: Ah I see. Thank you!

File: linera-core/src/unit_tests/wasm_worker_tests.rs [hunk 2]
```diff
@@ -246,7 +246,9 @@ where
         timestamp: Timestamp::from(2),
         ..make_state(Epoch::ZERO, creator_chain, admin_id)
     };
-    let creator_state = ExecutionStateView::from_system_state(creator_system_state.clone()).await;
+    let creator_state =
+        ExecutionStateView::from_system_state(creator_system_state.clone(), Default::default())
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-core/src/unit_tests/wasm_worker_tests.rs:250]: Haha, sure I can do that. I was historically making the same comment on people's PRs :)
In this case, I was a bit lazy and happy to not add the `use` import in many files.

File: linera-execution/src/runtime_actor/handlers.rs [hunk 3]
```diff
@@ -95,16 +95,18 @@ where
...
                 response_sender,
-            } => response_sender.respond(self.write().await.set_remaining_fuel(remaining_fuel)),
+            } => {
+                response_sender.respond(self.write().await.set_remaining_fuel(remaining_fuel).await)
+            }
             ContractRequest::TryReadAndLockMyState { response_sender } => response_sender.respond(
                 match self.write().await.try_read_and_lock_my_state().await {
                     Ok(bytes) => Some(bytes),
-                    Err(ExecutionError::ViewError(ViewError::NotFound(_))) => None,
+                    Err(ExecutionError::ViewError(_)) => None,
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-execution/src/runtime_actor/handlers.rs:109]: oops, I meant to undo this change.

File: linera-execution/src/resources.rs [hunk 4]
```diff
@@ -142,3 +142,54 @@ pub struct RuntimeCounts {
...
+        &mut self,
+        limits: &RuntimeLimits,
+        increment: u64,
+    ) -> Result<(), ExecutionError> {
+        if increment >= u64::MAX / 2 {
+            return Err(ExecutionError::ExcessiveRead);
+        }
+        self.bytes_read += increment;
+        if self.bytes_read >= limits.max_budget_bytes_read {
+            return Err(ExecutionError::ArithmeticError(ArithmeticError::Overflow));
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-execution/src/resources.rs:165]: Mmm this is weird. I thought I copied existing code. Let me fix.

File: linera-execution/src/lib.rs [hunk 5]
```diff
@@ -98,28 +113,35 @@ pub enum ExecutionError {
...
+        owned_by: Box<UserApplicationId>,
+    },
+    #[error("Session {0} is still opened at the end of a transaction")]
+    SessionWasNotClosed(SessionId),
+
     #[error("Attempted to call an application while the state is locked")]
-    ApplicationIsInUse,
-    #[error("Attempted to get an entry that is not locked")]
-    ApplicationStateNotLocked,
+    ApplicationIsInUse(UserApplicationId),
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-execution/src/lib.rs:139]: right

File: linera-execution/src/sync_runtime.rs [hunk 6]
```diff
@@ -0,0 +1,1054 @@
...
+    find_keys_queries: QueryManager<Keys>,
+    /// The find-key-values queries in progress.
+    find_key_values_queries: QueryManager<KeyValues>,
+}
+
+impl ViewUserState {
+    fn force_all_pending_queries(&mut self) -> Result<(), ExecutionError> {
+        self.read_value_queries.force_all()?;
+        self.find_keys_queries.force_all()?;
+        self.find_key_values_queries.force_all()?;
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-execution/src/sync_runtime.rs:182]: oh noo. That was a bug. Thanks!

File: linera-execution/src/sync_runtime.rs [hunk 7]
```diff
@@ -0,0 +1,1054 @@
...
+        state.pending_query = Some(receiver);
+        Ok(())
+    }
+
+    fn try_read_my_state_wait(&mut self, _promise: &Self::Read) -> Result<Vec<u8>, ExecutionError> {
+        let id = self.application_id()?;
+        let state = self
+            .simple_user_states
+            .get_mut(&id)
+            .ok_or(ExecutionError::PolledTwice)?;
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-execution/src/sync_runtime.rs:571]: Mmm there is no session here

**Review by jvff — APPROVED** [review]
Looks good! Feel free to land it, we can discuss if it's worth improving the `PolledTwice` error later. It might not even be needed if we make the promise APIs stateless :crossed_fingers:

File: linera-execution/src/sync_runtime.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-execution/src/sync_runtime.rs:571]: The promise is `()` here.

[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-execution/src/sync_runtime.rs]: Anyways, I renamed `PolledTwice` into `InvalidPromise` for this runtime (it's indeed more correct).

**Pull request merged**